contracts/extensions/ERC4907A.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+// ERC721A Contracts v4.1.0
+// Creator: Chiru Labs
+
+pragma solidity ^0.8.4;
+
+import './IERC4907A.sol';
+import '../ERC721A.sol';
+
+/**
+ * @title ERC4907A
+ *
+ * @dev [ERC4907](https://eips.ethereum.org/EIPS/eip-4907) compliant
+ * extension of ERC721A, which allows owners and authorized addresses
+ * to add a time-limited role with restricted permissions to ERC721 tokens.
+ */
+abstract contract ERC4907A is ERC721A, IERC4907A {
+    // The bit position of `expires` in packed user info.
+    uint256 private constant _BITPOS_EXPIRES = 160;
+
+    // Mapping from token ID to user info.
+    //
+    // Bits Layout:
+    // - [0..159]   `user`
+    // - [160..223] `expires`
+    mapping(uint256 => uint256) private _packedUserInfo;
+
+    /**
+     * @dev Sets the `user` and `expires` for `tokenId`.
+     * The zero address indicates there is no user.
+     *
+     * Requirements:
+     *
+     * - The caller must own `tokenId` or be an approved operator.
+     */
+    function setUser(
+        uint256 tokenId,
+        address user,
+        uint64 expires
+    ) public virtual {
+        address owner = ownerOf(tokenId);
+        if (_msgSenderERC721A() != owner)
```

Comment:
Let's add a comment here mentioning the gas savings of not having branching similar to what we have in 721A

Reply:
The if-if-if will branch.

The innards of ERC721A uses assembly to compute `msg.sender == owner || msg.sender == approvedAddress` branchlessly.

Reply:
What's the reason behind just making it 3 if statements instead of a single if?

Reply:
The following actually costs more gas 

```Solidity
if (_msgSenderERC721A() == owner || 
    isApprovedForAll(owner, _msgSenderERC721A() ||
    getApproved(tokenId) != _msgSenderERC721A())
```